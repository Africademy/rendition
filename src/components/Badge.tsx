import ColorHash from 'color-hash';
import assign from 'lodash/assign';
import memoize from 'lodash/memoize';
import * as React from 'react';
import styled from 'styled-components';
import asRendition from '../asRendition';
import { Coloring, RenditionSystemProps, Theme } from '../common-types';
import { getColor, px } from '../utils';
import Txt, { TxtProps } from './Txt';

const colorHash = new ColorHash();
const backgroundColor = memoize(
	(text: string): string => {
		return colorHash.hex(text.replace(/\s/g, ''));
	},
);

const BaseBadge = styled(Txt)`
	border-radius: ${props => px(props.theme.radius)};
	display: inline-block;
	min-width: 40px;
	text-align: center;
	font-weight: 800;
	font-style: normal;
	font-stretch: normal;
	letter-spacing: 0.5px;
`;

const Badge = ({
	xsmall,
	small,
	children,
	theme,
	...props
}: ThemedBadgeProps) => {
	let fontSize = 2;
	if (small) {
		fontSize = 1;
	}
	if (xsmall) {
		fontSize = 10;
	}

	if (typeof children !== 'string') {
		throw new Error(
			`The child element of the Badge component must be a string, received: ${typeof children}`,
		);
	}

	return (
		<BaseBadge
			p="2px 5px"
			fontSize={fontSize}
			{...props}
			color={props.color || '#fff'}
			bg={
				getColor(assign(props, { theme }), 'bg', 'main') ||
				backgroundColor(children)
			}
		>
			{children}
		</BaseBadge>
	);
};

export interface InternalBadgeProps extends TxtProps, Coloring {
	small?: boolean;
	xsmall?: boolean;
	color?: string;
	bg?: string;
}

export interface ThemedBadgeProps extends InternalBadgeProps {
	theme: Theme;
}

export type BadgeProps = InternalBadgeProps & RenditionSystemProps;

export default asRendition<React.FunctionComponent<BadgeProps>>(
	Badge,
	[],
	['bg', 'color'],
);
