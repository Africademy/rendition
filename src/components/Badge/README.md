# Badge

Badges should be used to present data that is not generated by the user (otherwise, tags should be used).
The badge can use any of the pre-defined shades, from which the background and text color are generated.

[View story source](https://github.com/balena-io-modules/rendition/blob/master/src/components/Badge/story.js)

## Props

| Name   | Type   | Default   | Required   | Description   |
| ------ | ------ | --------- | ---------- | ------------- |
| `shade`        | `number` | - | - | The index for one of the pre-determined shades for the badge |
| `children`  | `string` | - | âœ“ | The text to display inside the badge |
