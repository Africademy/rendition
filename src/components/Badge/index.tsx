import * as React from 'react';
import styled from 'styled-components';
import asRendition from '../../asRendition';
import { Coloring, RenditionSystemProps, Theme } from '../../common-types';
import { getLegibleTextColor } from '../../utils';
import Txt from '../Txt';

// All of the shades are checked so background to text contrast ratio is at least 4:1.
const shades = [
	'#C7FFFF',
	'#BAECC2',
	'#FFD4F1',
	'#F1CDE3',
	'#D7E4FE',
	'#FFD5D2',
	'#CAE1AC',
	'#D6D3FA',
	'#C3C7CA',
	'#FFDA93',
	'#E1EEE5',
	'#DBE0E3',
	'#FEEC63',
	'#FBE6F4',
	'#FFECAE',
	'#F1CDFF',
	'#C3EFFF',
	'#F3E8FD',
	'#E0FEB4',
	'#FDE8E7',
	'#CAFECD',
	'#D7E46C',
	'#F7D8BA',
];

const BaseBadge = styled(Txt.span)`
	font-family: ${props => props.theme.titleFont};
	display: inline-block;
	border-radius: 1em;
	line-height: 1;
`;

const Badge = ({
	children,
	className,
	shade,
	...otherProps
}: ThemedBadgeProps) => {
	if (typeof children !== 'string') {
		throw new Error(
			`The child element of the Badge component must be a string, received: ${typeof children}`,
		);
	}

	const shadeHex = shades[(shade || 0) % shades.length];

	return (
		<BaseBadge
			className={className}
			py={1}
			px="12px"
			fontSize={0}
			bg={shadeHex}
			{...otherProps}
			color={getLegibleTextColor(shadeHex)}
		>
			{children}
		</BaseBadge>
	);
};

export interface InternalBadgeProps extends Coloring {
	children: string;
	className?: string;
	shade?: number;
}

export interface ThemedBadgeProps extends InternalBadgeProps {
	theme: Theme;
}

export type BadgeProps = InternalBadgeProps & RenditionSystemProps;

export default asRendition<React.FunctionComponent<BadgeProps>>(Badge);
