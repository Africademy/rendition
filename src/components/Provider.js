import * as React from 'react'
import styled, { ThemeProvider } from 'styled-components'
import defaultTheme from '../theme'

const Base = styled.div`
  font-family: ${props => props.theme.font};
`

const Provider = ({ theme, ...props }) => {
  return (
    <ThemeProvider theme={theme || defaultTheme}>
      <Base {...props} />
    </ThemeProvider>
  )
}

export default Provider
