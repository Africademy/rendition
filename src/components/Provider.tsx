import * as React from 'react';
import styled, { StyledFunction, ThemeProvider } from 'styled-components';
import { DefaultProps, Theme } from '../common-types';
import defaultTheme from '../theme';

export interface ThemedBaseProps extends DefaultProps {
	theme?: Theme;
}

const Base = (styled.div as StyledFunction<DefaultProps>)`
  font-family: ${props => props.theme.font};
`;

const Provider = ({ theme, ...props }: ThemedBaseProps) => {
	return (
		<ThemeProvider theme={theme || defaultTheme}>
			<Base {...props} />
		</ThemeProvider>
	);
};

export default Provider;
