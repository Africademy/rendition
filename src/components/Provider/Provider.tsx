import {
	Grommet
} from 'grommet'
import * as React from 'react';

const Provider = ({ theme, ...props }: any) => {
	return (
		<Grommet theme={theme} {...props} />
	);
};

export default Provider;
