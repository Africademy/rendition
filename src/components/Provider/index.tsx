import { Grommet } from 'grommet';
import merge from 'lodash/merge';
import * as React from 'react';
import styled from 'styled-components';
import { DefaultProps, Theme } from '../../common-types';
import defaultTheme from '../../theme';
import { px } from '../../utils';
import { BreakpointProvider } from './BreakpointProvider';

const Base = styled(Grommet)`
	font-family: ${props => props.theme.font};
	font-size: ${props => px(props.theme.fontSizes[1])};

	h1,
	h2,
	h3,
	h4,
	h5,
	h6,
	button {
		font-family: ${props => props.theme.titleFont};
	}
`;

const Provider = ({ theme, ...props }: ThemedProvider) => {
	const providerTheme = merge(defaultTheme, theme);
	return (
		<BreakpointProvider breakpoints={providerTheme.breakpoints}>
			<Base theme={providerTheme} {...props} />
		</BreakpointProvider>
	);
};

export interface ThemedProvider extends DefaultProps {
	theme?: Partial<Theme>;
}

export default Provider;
