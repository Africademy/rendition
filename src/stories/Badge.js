import * as React from 'react'
import { storiesOf } from '@storybook/react'
import withReadme from 'storybook-readme/with-readme'
import styled from 'styled-components'
import { Badge } from '../'
import * as Readme from './README/Badge.md'

const Container = styled.div`
  margin: 30px;
`

storiesOf('Core/Badge', module)
  .addDecorator(withReadme(Readme))
  .addDecorator(story => <Container>{story()}</Container>)
  .add('Standard', () => {
    return (
      <React.Fragment>
        <Badge text='badge1' mr={2} />
        <Badge text='badge2' mr={2} />
        <Badge text='badge3' />
      </React.Fragment>
    )
  })
  .add('Controlled', () => {
    return (
      <React.Fragment>
        <Badge primary text='Primary' mr={2} />
        <Badge secondary text='Secondary' mr={2} />
        <Badge tertiary text='Tertiary' mr={2} />
        <Badge bg='purple' color='green' text='Custom' />
      </React.Fragment>
    )
  })
