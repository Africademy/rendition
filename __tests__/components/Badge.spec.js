/* globals expect, test */
import React from 'react'
import renderer from 'react-test-renderer'
import Provider from '../../src/components/Provider'
import Badge from '../../src/components/Badge'

test('Badge renders correctly', () => {
  const component = renderer.create(
    <Provider>
      <Badge>Badge</Badge>
    </Provider>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
