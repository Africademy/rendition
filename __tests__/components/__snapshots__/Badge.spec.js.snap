// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge renders correctly 1`] = `
<div
  className="StyledGrommet-sc-19lkkz7-0 bBubtL"
>
  <div
    className="sc-bdVaJa lgVSbN"
  >
    <div
      className="sc-gqjmRU juqtoI sc-gZMcBi bhtlIA sc-htoDjs giRJYk sc-gzVnrw jCPitn"
    >
      Badge
    </div>
  </div>
</div>
`;
